api/hasura/actions/_handlers/backfillPgive.ts

```diff
@@ -0,0 +1,418 @@
+            (Total Epoch pGIVE excluding Opt Out Bonus if any) * (% of GIVE received) */
+
+            recipients.forEach(recipient => {
+              if (recipient.gives_received) {
+                recipient.pgive =
+                  (recipient.gives_received / totalTokensSent) *
+                  (potTotal - totalOptOutPgiveAlloc);
+              }
+              recipient.pgive += recipient.opt_out_bonus;
+            });
+          }
+
+          // epoch insert objects
+          epochObjects.push({
+            pgive: roundNumbers(potTotal),
+            gives_receiver_base: roundNumbers(epoch.gives_receiver_base),
+            active_months_bonus: roundNumbers(epoch.active_months_bonus),
+            notes_bonus: roundNumbers(epoch.notes_bonus),
+            epoch_id: epoch.id,
+            active_months: epochIndexedData.activeMonths,
+          });
+        });
+
+        /* Normalize the epochs pgives in the same month for the recipients */
+
+        epochIndexedData.epochs.forEach(epoch => {
+          Object.keys(epochBasedData[epoch.id]).forEach(recipientId => {
+            const recipient = epochBasedData[epoch.id][recipientId];
+            if (epochIndexedData.epochs.length > 1) {
+              let epochsTotalPGive = 0;
+              epoch.normalizedEpochs.forEach(eId => {
+                epochsTotalPGive +=
+                  recipientId in epochBasedData[eId]
+                    ? epochBasedData[eId][recipientId].pgive
+                    : 0;
+              });
+
+              /* Epochs are normalized by simply dividing the total pGives and the number of epochs */
+
+              recipient.normalized_pgive =
+                epochsTotalPGive / epoch.normalizedEpochs.length;
+            } else {
+              recipient.normalized_pgive = recipient.pgive;
+            }
+
+            memberObjects.push({
+              pgive: roundNumbers(recipient.pgive),
+              user_id: recipient.user_id,
+              normalized_pgive: roundNumbers(recipient.normalized_pgive),
+              gives_received: recipient.gives_received,
+              opt_out_bonus: roundNumbers(recipient.opt_out_bonus),
+              epoch_id: epoch.id,
+            });
+          });
+        });
+      });
+
+      if (epochObjects.length) {
+        ops[`u${circle.id}_epoch_data`] = {
+          insert_epoch_pgive_data: [
+            {
+              objects: epochObjects,
+            },
+            { __typename: true },
+          ],
+        };
+      }
+    }
+  });
+
+  if (memberObjects.length) {
+    ops[`member_data`] = {
+      insert_member_epoch_pgives: [
+        {
+          objects: memberObjects,
+        },
+        { __typename: true },
+      ],
+    };
+  }
+  if (Object.keys(ops).length) {
+    await adminClient.mutate(
+      {
+        __alias: {
+          ...ops,
+        },
+      },
+      {
+        operationName: 'insert_pGiveData',
+      }
+    );
+  }
+  res.status(200).json({ success: true });
```

Comment:
This should return something like a count processed or something so you know when it's done or there are no more circles to process, otherwise how do you know when to stop backfilling? Also, how is this really going to be used? Just call in a loop until all circles have been handled? One concern is if a single circle takes >30s vercel will time us out.

Reply:
So i've moved this to a cron fully and batched it, the cron i've set it to timeout `120` seconds.
I'd tested the previous verions of the code with prod data before and the handler could run like 15-150 circles (depending the length of epochs it had) in 1 go . 

I've also added a `PGIVE_CIRCLE_MAX_PER_CRON` env var and defaulted it to 10 circles.

Theres a sub filter that filters away circles that already had pgive records generated , so the cron is going to keep processing until this filters to nothing.
```
circles: [
      {
        limit: parseInt(process.env.PGIVE_CIRCLE_MAX_PER_CRON || '') || 10,
        where: {
          epochs: {
            ended: { _eq: true },
            end_date: { _gte: startFrom.toISO(), _lte: endTo.toISO() },
            _not: {
              /* this filters away epochs that already has pgive data generated */
              pgive_data: {},
            },
          },
        },
      },
      {
        id: true,
      },
    ],
```